### DISCUSSION TITLE
MQTT Support for homeassistant integration

### DISCUSSION THREAD

**Discussion by ankurvdev** [gtxaspec/wz_mini_hacks#75] [General]
What would it take to add MQTT support 

Publishing and subscribing to MQTT bindings will make integration with Homeassistant really useful and eliminate need for frigate or any other host-based accelerated AI detection.

Looking at something similiar in Dafang-Hacks 
It seems they have a binary reading state from shared memory that fires off a shell script on motion detection

https://github.com/EliasKotlyar/Xiaomi-Dafang-Software/blob/1cfed9e1735c3bd376f86195318a9067161cb3d3/v4l2rtspserver-master/src/ImpEncoder.cpp#L723

**Comment by gtxaspec** [comment]
will need some help on this one...

**Reply by guarddog13** [reply]
Could you add curl support?

**Comment by Vendo232** [comment]
I vote for MQTT, it would be very cool feature which would turn V3 into sensor device.

**Comment by gtxaspec** [comment]
With the latest build, once you enable rtsp, ```cmd``` becomes available:

```cmd waitmotion <timeout>```

so, cmd waitmotion 1000 waits for motion, for 1000 seconds, then exits once motion is detected.  it outputs the coordinates of the motion detected. 

you could so something like:
```
#!/bin/sh

while true
do cmd waitmotion 1000
###Do whatever you want here
echo "Motion Detected"
cmd jpeg | tail -n +4 > /media/mmc/test.jpg
cmd aplay /usr/share/notify/binbin.wav 40
sleep 1
done
```
To get continuous output for use in a script.  This example prints text, takes a high-quality photo and plays a chime sound on the camera. 

Checkout the sources in libcallback if you want more info on how it works.

work in progress...

**Comment by Vendo232** [comment]
here is an idea: Execute script which will play certain *.wav audio file using shell command. 
use case: When alarm is triggered play script "intruder" with intruder.wav or if cam is in garage play script "Garage is closing clear the space... " 

in Home Assistant there is an option to execute scripts via shell_command function. It is possible to SSH to external shell prompts from Home Assistant. You can then SSH to any remote machine and perform commands.

in example execution of maintanance on remote IP:

`shell_command:
  raspi_maintenance: ssh -i /config/ssh/id_rsa -o 'StrictHostKeyChecking=no' username@remote_IP '~/sys_maintenance'
 
script:
 system_maintenance_raspi:
   alias: Raspberry Pi System Maintenance
   sequence:
     service: shell_command.raspi_maintenance`

then I can create simple automation to run this script when : alarm is triggered or garage is opened. 

setting up the SSH rsa keys is a hassle and I was unable to do it. Wanted to read CPU temp from another Debian server and pull the data to Home Assistant. 

How to setup SSH keys

https://siytek.com/home-assistant-shell/

will try it later this week on V3 and run audio_test.sh via SSH command from HA

#!/bin/sh
cmd aplay /usr/share/notify/swap.wav 40

**Comment by Matthew-Whitlock** [comment]
A workaround to full MQTT support is to use the webhook trigger in homeassistant (which I think makes more sense anyway) - I figured I'd post it here, as it took a little tinkering:

```
#!/opt/wz_mini/bin/bash

while true
do
  ret=`cmd waitmotion 1000 | grep "timeout" | wc -l`

  if [ $ret -eq 0  ] ; then
    echo "Reporting motion!"
    wget --post-data="motion=1" http://your.homeassistant.com/api/webhook/your_webhook -O/dev/null 2>/dev/null
  else
    #optionally actively report timeouts
    echo "No motion found!"
  fi
done
```

An important note is that the version of wget provided doesn't support https, so you'll have to expose http access if you aren't now. The post-data isn't important, and you can send any number of var1=data&var2=data2 within, which can be accessed from homeassistant.

**Reply by guarddog13** [reply]
`#!/opt/wz_mini/bin/bash

while true
do
  ret='cmd waitmotion 1000 | grep "timeout" | wc -l'

  if  $ret -eq 0 ; then
    echo "Reporting motion!"
    wget --post-data "motion=1" http://your.homeassistant.com/api/webhook/your_webhook -O/dev/null 2>/dev/null
  else
    #optionally actively report timeouts
    echo "No motion found!"
  fi
done`

I for the life of me couldn't get it to work until I changed a couple of things. Working beatifully now. Thank you for guiding me in the right direction.

**Comment by gtxaspec** [comment]
i'll try to compile a static version of wget with https, i forgot about that.

**Comment by gtxaspec** [comment]
wget with https support will be included in next release!

**Comment by gtxaspec** [comment]
release out now: ```/opt/wz_mini/bin/wget```

**Comment by gtxaspec** [comment]
just FYI, if you want `cmd waitMotion` to work, you need to enable "Motion Tagging" in the app on your phone for that specific camera.

**Reply by guarddog13** [reply]
Is there a list of the the available cmds? Like is there a cmd waitPerson if you're a plus member?

**Reply by gtxaspec** [reply]
not yet, coming soon if we can figure it out lol

you can type `cmd` for a partial list. check the read me for everything

**Reply by guarddog13** [reply]
If you could get waitPerson that would be amazing...that could effectively eliminate 90 percent of my need for the NVR as I don't mind recording on the SD cards. I would still need the NVR for facial recognition at the front door....now if you could get wait 'friendly face' it could make these $35  cameras worth more than the nest cams I'm working on replacing them with...

**Comment by mwhdc** [comment]
Experimenting with this - currently cmd waitMotion 600 (or any integer >= the minimum of 10) doesn't appear to wait the specified number of seconds and then timeout, instead it checks for motion once, issues "clear" if none, and exits. I have motion tagging enabled on the test camera. Did this behavior change in mmakada's code? The last commit was May 9.

**Reply by gtxaspec** [reply]
@mwhdc I believe this is fixed this in libcallback from my testing, to be included in the upcoming wz_mini release

**Reply by mwhdc** [reply]
> @mwhdc I believe this is fixed this in libcallback from my testing, to be included in the upcoming wz_mini release

Thanks, I ran the upgrade script on two v2 cams and one v3 this morning. cmd waitmotion seems (I haven't timed it) to handle the timeout correctly on the v3, but behavior on the v2s is unchanged. What's the best way to check that the v2 cams are actually running the 7/13 code? You might want to start bundling up batches of fixes and enhancements into versioned releases, and write a version.txt somewhere in /opt/wz_mini.

**Comment by g8keeperzuul** [comment]
MQTT with Home Assistant discovery layered on top of wz mini hacks:
https://github.com/g8keeperzuul/wz_mini_hacks-MQTT

**Reply by Vendo232** [reply]
it looks awesome! will give it a try later today

**Reply by baudev** [reply]
Thanks for your work! I'll look at it soon

**Reply by simonsickle** [reply]
I have sent a PR for motion detection sensors to be sent via MQTT. It works beautify in home assistant and lets me see when my baby is awake. https://github.com/g8keeperzuul/wz_mini_hacks-MQTT/issues/3

The prerequisite is to have RTSP enabled, so that the waitMotion command will work. It has a pretty quick cool down right now before it goes back to the "no motion" state. 


I'm sure this would work with Scrypted to notify HKSV of motion (to trigger recordings) but I've not played with that yet, I just have it added as a feed for the PiP on my Apple TV.

I'd be willing to port this into the main repo here if contributors were open to it and @g8keeperzuul was ok with me using his work to do so.

**Reply by gtxaspec** [reply]
@g8keeperzuul @simonsickle would love to see this in wz_mini, if you want to send a PR, i'm all for it!

**Comment by mach7** [comment]
@gtxaspec has this been brought into the core of wz_mini?

**Reply by gtxaspec** [reply]
just saw this, i'll take a look soon.  thanks for the ping.

**Comment by bnevets27** [comment]
@gtxaspec This would be a really great feature have pulled into wz_mini. Please take a look if you get a chance. Thanks!
